File: python/compresr/agents/tools/_agentcore.py
"""AgentCore Web Search client (Amazon Bedrock AgentCore over MCP).

Ported from the standalone ``websearch_client.py`` reference. Wraps the
Cognito client-credentials OAuth handshake plus an MCP streamable-HTTP session
into a synchronous :meth:`AgentCoreClient.search` call that works in plain
scripts and in Jupyter (handles the already-running event-loop case).

The runtime dependencies (``mcp``, ``nest-asyncio``) are **optional** — they are
imported lazily inside the call so a bare ``pip install compresr`` still imports
fine. The error only surfaces when ``WebSearchTool.agentcore(...)`` is actually
invoked without the extra installed::

    pip install compresr[agentcore]

Never logs the Cognito client secret or the bearer token.
"""

from __future__ import annotations

import asyncio
import json
import threading
from dataclasses import dataclass
from typing import Any, Optional

import httpx

# Guard against an unbounded/misbehaving gateway response before buffering it
# in memory and parsing as JSON.
MAX_RESPONSE_BYTES = 1_000_000

# The gateway also exposes this built-in *tool*-search (semantic search over the
# gateway's own tools). It is NOT web search and must be ignored.
TOOL_SEARCH_META = "x_amz_bedrock_agentcore_search"

TOKEN_TIMEOUT_SECONDS = 30

# AgentCore web search accepts maxResults in the inclusive range 1..25.
MIN_RESULTS = 1
MAX_RESULTS = 25

_MISSING_DEP_MSG = (
    "WebSearchTool.agentcore requires the 'agentcore' extra. "
    "Install: pip install compresr[agentcore]"
)


@dataclass(frozen=True)
class AgentCoreConfig:
    """Immutable connection config for the AgentCore web-search gateway."""

    gateway_url: str
    cognito_token_url: str
    client_id: str
    client_secret: str
    scope: str
    max_results: int = 5


def _clamp_results(value: int) -> int:
    """Clamp a requested result count to AgentCore's inclusive 1..25 range."""
    return max(MIN_RESULTS, min(MAX_RESULTS, value))


def _run(coro: Any) -> Any:
    """Run a coroutine whether or not an event loop is already running (Jupyter)."""
    try:
        loop = asyncio.get_running_loop()
    except RuntimeError:
        return asyncio.run(coro)
    # Inside a running loop (e.g. Jupyter): allow nested execution.
    try:
        import nest_asyncio
    except ImportError as exc:  # pragma: no cover - exercised via search()
        raise ImportError(_MISSING_DEP_MSG) from exc

    nest_asyncio.apply()
    return loop.run_until_complete(coro)


class AgentCoreClient:
    """Synchronous client for AgentCore web search over MCP.

    Mints (and caches) a Cognito bearer token, then opens an MCP
    streamable-HTTP session to call the gateway's web-search tool. A single
    instance is shared by the built tool so the token is reused across calls.
    """

    def __init__(self, config: AgentCoreConfig) -> None:
        self._config = config
        self._token: Optional[str] = None
        self._token_lock = threading.Lock()

    # --- auth ----------------------------------------------------------------

    def token(self, refresh: bool = False) -> str:
        """Return a cached bearer token, minting a fresh one when needed.

        Guarded by a lock (double-checked locking) so concurrent first-use
        calls from multiple threads only mint one token. The client secret
        and the resulting token are never logged.
        """
        if self._token and not refresh:
            return self._token
        with self._token_lock:
            # Re-check after acquiring the lock: another thread may have
            # already minted/refreshed the token while we were waiting.
            if self._token and not refresh:
                return self._token
            try:
                resp = httpx.post(
                    self._config.cognito_token_url,
                    data={
                        "grant_type": "client_credentials",
                        "client_id": self._config.client_id,
                        "client_secret": self._config.client_secret,
                        "scope": self._config.scope,
                    },
                    headers={"Content-Type": "application/x-www-form-urlencoded"},
                    timeout=TOKEN_TIMEOUT_SECONDS,
                )
                resp.raise_for_status()
                self._token = resp.json()["access_token"]
            except httpx.HTTPError as exc:
                # Re-raise sanitized: the original exception's .request (and
                # this frame's locals) carry the client_secret, which we do
                # not want chained into error-monitoring tools (e.g. Sentry).
                status = getattr(getattr(exc, "response", None), "status_code", None)
                detail = f" (HTTP {status})" if status is not None else ""
                raise RuntimeError(f"AgentCore Cognito token request failed{detail}.") from None
            return self._token

    # --- tool selection ------------------------------------------------------

    @staticmethod
    def _pick_web_search_tool(names: list[str]) -> str:
        """Pick the web-search tool, ignoring the semantic tool-search meta tool."""
        candidates = [n for n in names if n != TOOL_SEARCH_META]
        for name in candidates:
            if "websearch" in name.lower().replace("-", "").replace("_", ""):
                return name
        if not candidates:
            raise RuntimeError(f"No web-search tool found among {names}")
        return candidates[0]

    # --- search --------------------------------------------------------------

    async def _search_async(self, query: str, max_results: int, *, refresh: bool) -> dict:
        # Lazy import so a bare install imports fine; only fails on actual use.
        try:
            from mcp import ClientSession
            from mcp.client.streamable_http import streamablehttp_client
        except ImportError as exc:
            raise ImportError(_MISSING_DEP_MSG) from exc

        headers = {"Authorization": f"Bearer {self.token(refresh=refresh)}"}
        async with streamablehttp_client(self._config.gateway_url, headers=headers) as (
            read,
            write,
            _,
        ):
            async with ClientSession(read, write) as session:
                await session.initialize()
                names = [t.name for t in (await session.list_tools()).tools]
                tool = self._pick_web_search_tool(names)
                result = await session.call_tool(tool, {"query": query, "maxResults": max_results})
                return _parse_tool_result(result)

    def search(self, query: str, max_results: Optional[int] = None) -> list[dict]:
        """Run a web search and return normalized result dicts.

        Each result is ``{"title", "url", "content", "published_date"}`` — the
        ``content`` key matches the SDK's ``_flatten_search_results`` flattener.
        Re-mints the bearer token once on an MCP 401 and retries.
        """
        count = _clamp_results(self._config.max_results if max_results is None else max_results)
        try:
            payload = _run(self._search_async(query, count, refresh=False))
        except Exception as exc:  # noqa: BLE001 - inspect for a 401 to retry once
            if _is_unauthorized(exc):
                payload = _run(self._search_async(query, count, refresh=True))
            else:
                raise
        results = payload.get("results", [])
        if not isinstance(results, list):
            raise RuntimeError(
                "AgentCore web-search returned an unexpected 'results' shape "
                f"(expected a list, got {type(results).__name__})."
            )
        out: list[dict] = []
        for item in results:
            if not isinstance(item, dict):
                continue
            out.append(
                {
                    "title": item.get("title"),
                    "url": item.get("url"),
                    "content": item.get("text", "") or "",
                    "published_date": item.get("publishedDate"),
                }
            )
        return out


def _parse_tool_result(result: Any) -> dict:
    """Validate and parse an MCP ``CallToolResult`` into a JSON dict.

    Defensive against everything a misbehaving/compromised gateway could
    return: a tool-level error (``isError=True`` with plain-text content,
    NOT JSON), an oversized body, malformed JSON, or a JSON value that isn't
    an object. Raises ``RuntimeError`` with a clear, secret-free message in
    every case instead of letting a bare ``JSONDecodeError``/``AttributeError``
    propagate.
    """
    text = "".join(getattr(b, "text", "") for b in result.content)
    if getattr(result, "isError", False):
        # Gateway tool-level errors (bad query, throttling, auth/scope error)
        # come back as plain text, NOT JSON. Surface them directly instead of
        # failing on json.loads with a confusing JSONDecodeError. The text may
        # itself contain "401" — keep it intact so the unauthorized-retry
        # heuristic in _is_unauthorized still matches.
        raise RuntimeError(f"AgentCore web-search tool error: {text}")
    if len(text.encode("utf-8", errors="ignore")) > MAX_RESPONSE_BYTES:
        raise RuntimeError(
            f"AgentCore web-search response exceeded the {MAX_RESPONSE_BYTES} byte size guard."
        )
    try:
        payload = json.loads(text)
    except json.JSONDecodeError as exc:
        raise RuntimeError(f"AgentCore web-search returned malformed JSON: {exc}") from None
    if not isinstance(payload, dict):
        raise RuntimeError(
            "AgentCore web-search returned an unexpected JSON shape "
            f"(expected an object, got {type(payload).__name__})."
        )
    return payload


def _is_unauthorized(exc: BaseException) -> bool:
    """Best-effort detection of an MCP/HTTP 401 to trigger a single token re-mint."""
    status = getattr(getattr(exc, "response", None), "status_code", None)
    if status == 401:
        return True
    return "401" in str(exc)


__all__ = ["AgentCoreConfig", "AgentCoreClient"]

File: typescript/src/agents/tools/_agentcore.ts
/**
 * AgentCore web-search client — Cognito OAuth + MCP streamable-HTTP.
 *
 * Ported from the reference ``websearch_client.py`` (``AgentCoreWebSearch``).
 * Reaching Amazon Bedrock AgentCore web search requires two hops that the
 * off-the-shelf Tavily/Brave tools handle for us but AgentCore does not:
 *
 *   1. A Cognito **client-credentials** OAuth grant to mint a bearer token.
 *   2. An **MCP streamable-HTTP** session to the gateway, calling the
 *      ``...WebSearch`` tool with ``{query, maxResults}`` and parsing the
 *      single JSON text block it returns.
 *
 * ``@modelcontextprotocol/sdk`` is an **optional** peer: a bare install must
 * import fine; the ``missing_peer_dependency`` error only fires when
 * ``.agentcore()`` is actually invoked. Never log the client secret or the
 * bearer token.
 *
 * Mirrors Python ``compresr/agents/tools/_agentcore.py``.
 */
import { CompresrError } from '../../errors/index.js';

/**
 * The gateway also exposes this built-in *tool*-search (semantic search over
 * the gateway's own tools). It is NOT web search — always ignore it.
 */
const TOOL_SEARCH_META = 'x_amz_bedrock_agentcore_search';

/** Cognito token POST timeout (ms). */
const TOKEN_TIMEOUT_MS = 30_000;

/** MCP tool-call timeout (ms) — bounds how long a hung gateway can stall a search. */
const TOOL_CALL_TIMEOUT_MS = 30_000;

/**
 * Upper bound (bytes, UTF-8) on the MCP tool-result text before it is
 * buffered and JSON.parse()'d. Guards against an unbounded/misbehaving
 * gateway response causing unbounded memory allocation. Mirrors Python's
 * ``MAX_RESPONSE_BYTES``.
 */
const MAX_RESPONSE_BYTES = 1_000_000;

/** Resolved AgentCore gateway configuration (camelCase). */
export interface AgentCoreConfig {
  gatewayUrl: string;
  cognitoTokenUrl: string;
  clientId: string;
  clientSecret: string;
  scope: string;
  /** Default result count; clamped 1–25 by the caller. */
  maxResults: number;
}

/** One web-search hit. ``content`` is AWS's extracted snippet, not full text. */
export interface AgentCoreSearchResult {
  title: string | null;
  url: string | null;
  content: string;
  publishedDate: string | null;
}

/** Minimal client surface — ``search`` mints/caches the token as needed. */
export interface AgentCoreClient {
  search(query: string, maxResults: number): Promise<AgentCoreSearchResult[]>;
}

// --- minimal structural types for the optional MCP SDK ---------------------
// The SDK is an optional peer, so we narrow to only the surface we use rather
// than depending on its exported types at compile time.

interface McpTransport {
  close(): Promise<void>;
}

interface McpTransportCtor {
  new (
    url: URL,
    opts?: {
      requestInit?: { headers?: Record<string, string>; signal?: AbortSignal };
    }
  ): McpTransport;
}

interface McpToolResult {
  content: unknown;
  /** Per the MCP spec, a failed tool call still returns HTTP 200 with
   * ``isError: true`` and the error message in the text block. */
  isError?: boolean;
}

interface McpClient {
  connect(transport: McpTransport): Promise<void>;
  listTools(): Promise<{ tools: Array<{ name: string }> }>;
  callTool(params: {
    name: string;
    arguments: Record<string, unknown>;
  }): Promise<McpToolResult>;
  close(): Promise<void>;
}

interface McpClientCtor {
  new (info: { name: string; version: string }): McpClient;
}

interface LoadedMcp {
  Client: McpClientCtor;
  StreamableHTTPClientTransport: McpTransportCtor;
}

/**
 * Dynamically import the optional ``@modelcontextprotocol/sdk``. The exact
 * client subpaths were verified against the installed package (v1.29);
 * ``./client/index.js`` and ``./client/streamableHttp.js`` resolve via the
 * package's ``./*`` export wildcard.
 */
async function loadMcp(): Promise<LoadedMcp> {
  let clientMod: Record<string, unknown>;
  let transportMod: Record<string, unknown>;
  try {
    clientMod = await import('@modelcontextprotocol/sdk/client/index.js');
    transportMod = await import(
      '@modelcontextprotocol/sdk/client/streamableHttp.js'
    );
  } catch {
    throw new CompresrError(
      "WebSearchTool.agentcore requires '@modelcontextprotocol/sdk'. " +
        'Install with: npm install @modelcontextprotocol/sdk',
      'missing_peer_dependency'
    );
  }
  const Client = clientMod['Client'] as McpClientCtor | undefined;
  const StreamableHTTPClientTransport = transportMod[
    'StreamableHTTPClientTransport'
  ] as McpTransportCtor | undefined;
  if (
    typeof Client !== 'function' ||
    typeof StreamableHTTPClientTransport !== 'function'
  ) {
    throw new CompresrError(
      "'@modelcontextprotocol/sdk' is missing expected client exports — " +
        'upgrade the package.',
      'missing_peer_dependency'
    );
  }
  return { Client, StreamableHTTPClientTransport };
}

/** Normalize a tool name for fuzzy matching (drop case, dashes, underscores). */
function normalizeName(name: string): string {
  return name.toLowerCase().replace(/[-_]/g, '');
}

/** Pick the web-search tool, ignoring the semantic tool-search meta tool. */
function pickWebSearchTool(names: ReadonlyArray<string>): string {
  const candidates = names.filter((n) => n !== TOOL_SEARCH_META);
  for (const name of candidates) {
    if (normalizeName(name).includes('websearch')) return name;
  }
  if (candidates.length === 0) {
    throw new CompresrError(
      `No web-search tool found among [${names.join(', ')}].`,
      'agentcore_no_tool'
    );
  }
  return candidates[0];
}

/** Extract the ``access_token`` string from an untrusted token response. */
function extractAccessToken(data: unknown): string | null {
  if (data && typeof data === 'object' && 'access_token' in data) {
    const token = (data as { access_token?: unknown }).access_token;
    if (typeof token === 'string' && token.length > 0) return token;
  }
  return null;
}

/** Concatenate the ``text`` of all text content blocks. */
function extractText(content: unknown): string {
  if (!Array.isArray(content)) return '';
  let out = '';
  for (const block of content) {
    if (block && typeof block === 'object' && 'text' in block) {
      const text = (block as { text?: unknown }).text;
      if (typeof text === 'string') out += text;
    }
  }
  return out;
}

/** Parse the gateway's JSON text block into normalized result rows. */
function parseResults(text: string): AgentCoreSearchResult[] {
  if (!text) return [];
  if (Buffer.byteLength(text, 'utf8') > MAX_RESPONSE_BYTES) {
    throw new CompresrError(
      'AgentCore web search response exceeded the size guard.',
      'agentcore_bad_response'
    );
  }
  let payload: unknown;
  try {
    payload = JSON.parse(text);
  } catch {
    throw new CompresrError(
      'AgentCore web search returned a non-JSON payload.',
      'agentcore_bad_response'
    );
  }
  if (!payload || typeof payload !== 'object' || !('results' in payload)) {
    return [];
  }
  const results = (payload as { results?: unknown }).results;
  if (!Array.isArray(results)) return [];
  const out: AgentCoreSearchResult[] = [];
  for (const item of results) {
    if (!item || typeof item !== 'object') continue;
    const row = item as Record<string, unknown>;
    out.push({
      title: typeof row['title'] === 'string' ? row['title'] : null,
      url: typeof row['url'] === 'string' ? row['url'] : null,
      content: typeof row['text'] === 'string' ? row['text'] : '',
      publishedDate:
        typeof row['publishedDate'] === 'string' ? row['publishedDate'] : null,
    });
  }
  return out;
}

/** Best-effort detection of an HTTP 401 / unauthorized failure. */
function isUnauthorized(err: unknown): boolean {
  if (err && typeof err === 'object') {
    const code = (err as { code?: unknown }).code;
    if (code === 401) return true;
    const message = (err as { message?: unknown }).message;
    if (typeof message === 'string') {
      const lowered = message.toLowerCase();
      return lowered.includes('401') || lowered.includes('unauthorized');
    }
  }
  return false;
}

/**
 * Build an {@link AgentCoreClient}. The bearer token is cached across calls and
 * transparently re-minted once on an MCP 401. Secrets and the token itself are
 * never logged or surfaced in error messages.
 */
export function buildAgentCoreClient(config: AgentCoreConfig): AgentCoreClient {
  let cachedToken: string | null = null;
  // Cache the in-flight mint promise (not just the resolved token) so
  // concurrent first-use callers await the same Cognito request instead of
  // each firing their own (redundant credential-bearing requests / self
  // -inflicted rate limiting).
  let inFlightMint: Promise<string> | null = null;

  async function mintTokenUncached(): Promise<string> {
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), TOKEN_TIMEOUT_MS);
    let resp: Response;
    try {
      resp = await fetch(config.cognitoTokenUrl, {
        method: 'POST',
        headers: { 'content-type': 'application/x-www-form-urlencoded' },
        body: new URLSearchParams({
          grant_type: 'client_credentials',
          client_id: config.clientId,
          client_secret: config.clientSecret,
          scope: config.scope,
        }),
        signal: controller.signal,
      });
    } catch (err: unknown) {
      // Do NOT rethrow the original error — it (and any .cause it carries)
      // may reference the request, including the client_secret in its body.
      // Covers both network-level failures (DNS/TLS/connection reset) and
      // the abort fired by the timeout above.
      const timedOut = err instanceof Error && err.name === 'AbortError';
      throw new CompresrError(
        timedOut
          ? 'AgentCore Cognito token request timed out.'
          : 'AgentCore Cognito token request failed (network error).',
        'agentcore_auth_error'
      );
    } finally {
      clearTimeout(timer);
    }
    if (!resp.ok) {
      // Do NOT include the response body — it may echo request credentials.
      throw new CompresrError(
        `AgentCore Cognito token request failed (HTTP ${resp.status}).`,
        'agentcore_auth_error'
      );
    }
    const data: unknown = await resp.json();
    const token = extractAccessToken(data);
    if (!token) {
      throw new CompresrError(
        'AgentCore Cognito response did not contain an access_token.',
        'agentcore_auth_error'
      );
    }
    cachedToken = token;
    return token;
  }

  async function mintToken(): Promise<string> {
    if (inFlightMint) return inFlightMint;
    const promise = mintTokenUncached().finally(() => {
      inFlightMint = null;
    });
    inFlightMint = promise;
    return promise;
  }

  async function getToken(refresh: boolean): Promise<string> {
    if (cachedToken && !refresh) return cachedToken;
    return mintToken();
  }

  async function callOnce(
    mcp: LoadedMcp,
    query: string,
    maxResults: number,
    refresh: boolean
  ): Promise<AgentCoreSearchResult[]> {
    const token = await getToken(refresh);
    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), TOOL_CALL_TIMEOUT_MS);
    const transport = new mcp.StreamableHTTPClientTransport(
      new URL(config.gatewayUrl),
      {
        requestInit: {
          headers: { Authorization: `Bearer ${token}` },
          signal: controller.signal,
        },
      }
    );
    const client = new mcp.Client({
      name: 'compresr-agentcore-client',
      version: '1.0.0',
    });
    try {
      await client.connect(transport);
      const listed = await client.listTools();
      const toolName = pickWebSearchTool(listed.tools.map((t) => t.name));
      const result = await client.callTool({
        name: toolName,
        arguments: { query, maxResults },
      });
      const text = extractText(result.content);
      if (result.isError) {
        // Gateway tool-level errors (bad query, throttling, auth/scope
        // error) come back as plain text, NOT JSON — surface the real
        // diagnostic text (and keep it intact for isUnauthorized() below).
        throw new CompresrError(
          `AgentCore web search tool returned an error: ${text}`,
          'agentcore_tool_error'
        );
      }
      return parseResults(text);
    } finally {
      clearTimeout(timer);
      await client.close().catch(() => undefined);
    }
  }

  async function search(
    query: string,
    maxResults: number
  ): Promise<AgentCoreSearchResult[]> {
    const mcp = await loadMcp();
    try {
      return await callOnce(mcp, query, maxResults, false);
    } catch (err: unknown) {
      if (isUnauthorized(err)) {
        // Token may have expired — re-mint once and retry.
        return callOnce(mcp, query, maxResults, true);
      }
      throw err;
    }
  }

  return { search };
}
